src/effects/effectchain.h

```diff
@@ -1,69 +1,85 @@
+/// startup and shutdown. EffectChain adds exactly one EngineEffectChain to
+/// the audio engine which is removed from the engine in the EffectChain
+/// destructor. This is in contrast to EffectSlot, which loads/unloads
+/// EngineEffects at the user's request.
+///
+/// The state of an EffectChain can be saved to and loaded from an
+/// EffectChainPreset, which can serialize/deserialize that state to/from XML.
 class EffectChain : public QObject {
     Q_OBJECT
   public:
-    EffectChain(EffectsManager* pEffectsManager, const QString& id,
-                EffectChainPointer prototype=EffectChainPointer());
+    EffectChain(const QString& group,
+            EffectsManager* pEffectsManager,
+            EffectsMessengerPointer pEffectsMessenger,
+            SignalProcessingStage stage = SignalProcessingStage::Postfader);
     virtual ~EffectChain();
 
-    void addToEngine(EngineEffectRack* pRack, int iIndex);
-    void removeFromEngine(EngineEffectRack* pRack, int iIndex);
-    void updateEngineState();
+    QString group() const;
 
-    // The ID of an EffectChain is a unique ID given to it to help associate it
-    // with the preset from which it was loaded.
-    const QString& id() const;
+    EffectSlotPointer getEffectSlot(unsigned int slotNumber);
 
-    // Whether the chain is enabled (eligible for processing).
-    bool enabled() const;
-    void setEnabled(bool enabled);
+    void registerInputChannel(const ChannelHandleAndGroup& handleGroup,
+            const double initialValue = 0.0);
+    /// Do not store this in EffectSlot! The enabled input channels are a property
+    /// of the chain, not the effect.
+    const QSet<ChannelHandleAndGroup>& getActiveChannels() const {
+        return m_enabledInputChannels;
+    }
 
-    // Activates EffectChain processing for the provided channel.
-    void enableForInputChannel(const ChannelHandleAndGroup& handleGroup);
-    bool enabledForChannel(const ChannelHandleAndGroup& handleGroup) const;
-    const QSet<ChannelHandleAndGroup>& enabledChannels() const;
-    void disableForInputChannel(const ChannelHandleAndGroup& handleGroup);
+    double getSuperParameter() const;
+    void setSuperParameter(double value, bool force = false);
+
+    EffectChainMixMode mixMode() const;
+    void setMixMode(EffectChainMixMode mixMode);
 
-    EffectChainPointer prototype() const;
+    const QString& getGroup() const {
+        return m_group;
+    }
 
-    // Get the human-readable name of the EffectChain
-    const QString& name() const;
-    void setName(const QString& name);
+    const QString& presetName() const;
+    void setPresetName(const QString& name);
 
     // Get the human-readable description of the EffectChain
     QString description() const;
     void setDescription(const QString& description);
 
-    double mix() const;
-    void setMix(const double& dMix);
-
     static QString mixModeToString(EffectChainMixMode type) {
```

Comment:
I'm not sure how to implement this better. I thought about moving the enum class to a new file and giving it to/from string methods, but C++ enum classes cannot have methods. I could make it a class, but then there still needs to be an enum class. Maybe I could make a namespace with a few stateless functions?

Reply:
My original idea was just to move it into a common file, in a way that a user of the enum also has access to the string representation. 

By the way: You can make a class with a public old enum. This is basically the same as a enum class, but can contain functions. The old enums are only dismissed, because they pollute the global namespace. This is not the case for enums wrapped in classes.

Reply:
Yes you can make a class with an old fashioned C enum but that does not have type safety.

@uklotzde do you have any suggestions about this?

Reply:
What kind of type safety are you referring to?

Reply:
C enums behave just like their base type. An int and an enum value can get mixed up and the compiler won't fail.

Reply:
d601fcb4d3bd7179e631367eaffb4fbd806f98e2